public/app/core/services/echo/consumers/PerformanceConsumer.ts

```diff
@@ -0,0 +1,74 @@
+import { getBackendSrv } from '@grafana/runtime';
+import { EchoConsumer, EchoEvent, EchoEventType } from '../types';
+import { Echo } from '../Echo';
+import { echoConsumerFactory } from '../echoConsumerFactory';
+
+export interface PerformanceEventPayload {
+  metricName: string;
+  duration: number;
+}
+
+export interface PerformanceEvent extends EchoEvent<EchoEventType.Performance, PerformanceEventPayload> {}
+
+export interface PerformanceConsumerOptions {
+  url: string;
+}
+
+/**
+ * Echo's performance metrics consumer
+ * Reports performance metrics to given url
+ */
+export class PerformanceConsumer implements EchoConsumer<PerformanceEvent, PerformanceConsumerOptions> {
+  private consumedPageLoadMetrics = false;
+  private buffer: PerformanceEvent[] = [];
+  supportedEvents = [EchoEventType.Performance];
+
+  constructor(private echoInstance: Echo, public options: PerformanceConsumerOptions) {}
+
+  consume = (e: EchoEvent) => {
+    if (this.supportedEvents.indexOf(e.type) > -1) {
+      this.echoInstance.logDebug('Performance consumer consumed: ', e);
+      this.buffer.push(e);
+    }
+  };
+
+  getPageLoadMetrics = () => {
+    const paintMetrics = performance.getEntriesByType('paint');
+    for (const metric of paintMetrics) {
+      this.buffer.push({
+        type: EchoEventType.Performance,
+        payload: {
+          metricName: metric.name,
+          duration: Math.round(metric.startTime + metric.duration),
+        },
+        meta: { ...this.echoInstance.getMeta() },
+        ts: performance.now(),
+      });
+    }
+  };
```

Comment:
Bit confused what a consumer is, though it only listened to events and published metrics to an endpoint? But it does collection as well?

Reply:
This is a specific one as paint metric is collected only once. If I were to collect this with [PerformanceObserver](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceObserver) I would have to initialise Echo as a very first thing on the page so that performance observer callback had access to the service already. I don't want to do this - it would require some major changes to our webpack setup, introducing new chunk etc. Hence this is how the PerformanceConsumer handles this metric consumption

Reply:
getPageLoadMetrics feels out of place in this class, which is about listening for events and buffering them, then writing to the http api. 

Still a bit confused, there is a buffer in this consumer and in the Echo service, both have a flush.  What is the purpose of each buffer / flush?

Reply:
The buffer in the EchoSrv could be skipped. it's not used at all (WIP).

`flush` in EchoSrv is only to have the iteration for consumer's `flush` invocation encapsulated in a function, that gets called on a given interval.

Also, I think the `consumer` is a wrong term here and we should use sth more like a `backend` maybe?